### ISSUE TITLE
NonZero* type constructors should be const fn

### ISSUE THREAD

**Issue by dhardy** [rust-lang/rust#58732]
E.g. `NonZeroU32::new_unchecked` is a `const fn` but `NonZeroU32::new` is not; this means that defining constant values is `unsafe`.
```rust
pub const X: NonZeroU32 = unsafe {
    NonZeroU32::new_unchecked(123)
};
```

**Comment by SimonSapin** [comment]
This depends on https://github.com/rust-lang/rust/issues/49146.

**Comment by m-ou-se** [comment]
This has been fixed in 1.47. `NonZero*::new` is const now.
